### PR TITLE
feat: enhance topic modeling with UMAP and HDBSCAN integration

### PR DESCRIPTION
**Pull request by sajz**
## Checklist

- [x] `poetry run pytest` passes locally with my changes
- [x] Tests have been added or updated to cover the changes
- [x] Documentation has been updated as needed
- [x] Code style checks pass
- [x] Commit messages follow the [Conventional Commits](https://www.conventionalcommits.org/) guidelines

## Description of Changes
- Updated `bert.py` to include UMAP and HDBSCAN for improved clustering in BERTopic.
  - Added UMAP with custom parameters for tighter clusters.
  - Integrated HDBSCAN with smaller minimum cluster size and prediction data.
  - Adjusted `n_gram_range` to allow up to trigrams and enabled probability calculation.
- Refactored main function to `fit_and_evaluate_model` for better modularity.
- Enhanced `concord.py` to reflect new model import and execution.
- Minor adjustments in preprocessing to streamline text normalization and improve clarity.

### CODE REVIEW

File: concord/concord.py [hunk 1]
```diff
@@ -4,75 +4,50 @@
...
-    """
-    Preprocess the documents by:
-    - Lowercasing
-    - Removing punctuation
-    - Removing stop words
-    - Lemmatizing
-    """
+    nltk.download('stopwords')
+    nltk.download('punkt')
+    nltk.download('wordnet')
```
Comment (Nic0-L) [inline concord/concord.py:13]: Manque un appel à nltk.download('punkt_tab') pour pouvoir utiiser wordnet

**Comment by Nic0-L** [comment]
Là où les topics étaient assez précis avant, maintenant parmis les 10 documents suivants (de la source fetch_20newsgroups) Bert en a catégorisé 5 dans le groupe "0_card_bus_irq_note". Sur l'échantillon de test de messages au sujet de l'impression 3D, Bert n'arrive même plus à trouver de topic. Il faut peut-être être moins aggressif sur les stopwords ?

Document-Topic Distribution:
                                             Document  Topic                                 Name
0   sure bashers pen fan pretty confused lack kind...      3                 3_dave_mtl_rick_mike
1   brother market video card support vesa local b...      0                  0_card_bus_irq_note
2   finally said dream mediterranean new area grea...      4        4_said_muslim_people_armenian
3   think scsi card dma transfer disk scsi card dm...      0                  0_card_bus_irq_note
4   old jasmine drive use new system understanding...      0                  0_card_bus_irq_note
..                                                ...    ...                                  ...
95  discussing use irq put little differently irq ...      0                  0_card_bus_irq_note
96  last question year notebook meg ram never prob...     -1           -1_sound_channel_one_thing
97  hey man spent past season learning skate playe...      3                 3_dave_mtl_rick_mike
98  anyone publically considered possibility fire ...      7  7_investigation_feustel_fire_caused
99                                opened distribution      5        5_window_app_application_apps

**Comment by Septimus4** [comment]
I have a branch"improve-accuracy", where I implemented a few technics from BerTopic doc, It would be great to test Saj branch rebased on mine since I also improved loggin to have more info, like weights per word in topics.

File: concord/concord.py (hunk 1 above)
Comment (Septimus4) [reply concord/concord.py:13]: This should not be there, this downloads wordlist basically, and I don't want our code to automatically and without user supervision download stuff on the user/dev computer, futhermore calling download blindly will setup the wordlist in the user home which is not great.
This should be handle by the developer directly and just provided to Concord by env variable like wrote in the readme file

**Review by Septimus4 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**